### ISSUE TITLE
Design of file creation RPC call

### ISSUE THREAD

**Issue by pkgw** [HERA-Team/librarian#2]
When you copy files from one Librarian to another, the copy task rsyncs the data, then uses the `create_file` RPC call to instruct the other Librarian to create a database record corresponding to the new file. However, it doesn't issue a `create_observation` call, so the other Librarian ends up with an inconsistent database lacking the observation metadata associated with those files.

This could of course be fixed by adding in an appropriate `create_observation` call in `copier`, but it seems to me that they should just be merged into one call.

```
register_file (store_name, jd, pol, djd, filename, filetype, size, md5)
```

For this to work, the Librarian needs to be smart enough not to add new rows if the files/observations have already been registered. In MySQL this can be done with `INSERT ... ON DUPLICATE KEY UPDATE` and I'm sure there's something analogous in Postgres. Stepping back, I think it's pretty clear having re-registration be a successful noop is the right semantics for the RPC call anyway.

**Comment by dannyjacobs** [comment]
If an observation has already been registered it should just add the new file to the list.  

Anderson and I had discussed at length the Problem of Apparently Identical Files. The correct thing to do is to give each file a unique hash and just add it to the list of files associated with the observation. However, how then does one distinguish between otherwise identical files? The answer we came up with was to track ingestion date, treating it like a backup archive or version control.  So anyway, duplicate files are fine and good, duplicate observations are evil.

In postgres I usually do a two step, SELECT and then either INSERT or UPDATE as needed. Something cleverer is always ok too.

**Comment by pkgw** [comment]
The SELECT-then-INSERT-or-UPDATE two-step is like, the *classic* technique that leads to race conditions and duplicated database rows! Wherever we write code like that it *will* come to bite us in the ass.

**Comment by pkgw** [comment]
Also, what do you mean by "give each file a unique hash"? Identical files will have identical MD5 sums by definition ...

**Comment by dannyjacobs** [comment]
Yeah, race conditions are a problem, but since we haven't yet had a mimo operating condition, I've left as an optimization to learn about later. If you know how do it, then go for it.

Should have said unique _key_ instead of hash.

**Comment by miguelfmorales** [comment]
If I’m following the discussion (I apologize if I’m way behind), I believe the key here is transactions — it is the way to combine a whole series of observations into one atomic transaction, and guarantees a consistent database. 

Peter’s original create_file issue may be a closely related issue of foreign keys, which could be avoided (?) by making sure there is an obsID foreign key. (Impossible to make a file row unless the observation exists).

Possible topics for Friday? 

M

**Comment by dannyjacobs** [comment]
files definitely need a foreign key to observations. If not, thats an oversight. 

I'll not be there on Friday, but I suggest that we put a thorough walkthrough of the Librarian and RTP at the start of our agenda for next week.

**Comment by pkgw** [comment]
FWIW: no, `file` does not have an `observation` foreign key right now: <https://github.com/HERA-Team/librarian/blob/master/hl_constraints.sql>

**Comment by pkgw** [comment]
Oversight corrected in [a8b932e](https://github.com/HERA-Team/librarian/commit/a8b932e5b6f2c626eb22c35946877e81088066fc).

**Comment by pkgw** [comment]
This has all been sorted out in the rewrite.
